Objects/longobject.c

```diff
@@ -64,6 +64,10 @@ get_small_int(sdigit ival)
     do if (-NSMALLNEGINTS <= ival && ival < NSMALLPOSINTS) { \
         return get_small_int((sdigit)ival); \
     } while(0)
+#define CHECK_SMALL_UINT(ival) \
+    do if (ival < NSMALLPOSINTS) { \
```

Comment:
Can you fix the comparison of signed with unsigned here? It may also be good to add a test for the case where `ival == (size_t)(-1)` (for example), just to make sure that the code does the right thing.

Reply:
> Can you fix the comparison of signed with unsigned here?

I'm not sure what should be fixed here, could you elaborate on that?

Reply:
I'm not experienced with the c-api, so take my interpretation with a grain of salt.

From my understanding, I think @mdickinson meant to cast `sdigit` to `ival` in the comparison, to ensure that two signed integers are being compared:
```c
#define CHECK_SMALL_UINT(ival) \
    do if (((sdigit)ival) < NSMALLPOSINTS) { \
```
Correct me if I'm mistaken.

Reply:
`ival` is unsigned and `NSMALLPOSINTS` is signed, but as far as I understand this doesn't produce any problems since `NSMALLPOSINTS` is non-negative.

Reply:
@sir-sigurd You're right, it should be the other way around, not sure why I wrote `ival` as signed.

Reply:
Indeed `NSMALLPOSINTS` is signed; `ival` is unsigned. My recollection is that some compilers will complain about comparisons between unsigned and signed types - in this case it's fine, but the C rules can give surprising results in other cases, so one often needs a cast to suppress a compiler warning.

But I may be misremembering: maybe compilers are smart enough to figure out that this isn't a case that needs a warning, given that `NSMALLPOSINTS` is a known-at-compile-time constant.

Reply:
I think compilers are smart enough not to complain about this, at least when the signed-but-positive value is a literal. Which is good, because there'd be a lot of noise if they did!

A super handy tool for experimenting with this kind of question is Compiler Explorer. Here's a test program:
https://godbolt.org/z/ZYt9pI

I've enabled every possible warning on clang, the relevant warning on GCC, and (I think) every possible warning on MSVC.

None of them complain when it's a literal in the right range, which is our case here.

For entertainment and wider context:
* All complain when it's an unknown value.
* All complain when it's 0 (so the comparison is always false).
* When it's a `const` value (rather than a literal) that is in the right range, clang accepts it but GCC and MSVC warn. You might change the constant's type, or add a cast.
* When it's a literal that is too big so the comparison is always true, clang and GCC warn but MSVC doesn't.

Reply:
Thanks for the detailed explanations, @mdickinson and @gnprice! I've recently been trying to help out with more PR reviews (since that seems to a bottleneck for CPython), but my experience is primarily with Python and writing documentation, not with C. I have some idea of the fundamentals, but I'm still learning.

I was misunderstanding the suggestion initially, and thought that the issue was that since `ival` is unsigned, it needed to be adjusted in value in order to accurately compare it against the constant `NSMALLPOSINTS`. So effectively, we would be converting `ival` into a signed integer just for the comparison. I suppose the correct way to do this would be adding a cast (if it were necessary, but it looks like it isn't in this case since `ival` is a literal).

>When it's a const value (rather than a literal) that is in the right range, clang accepts it but GCC and MSVC warn. You might change the constant's type, or add a cast.

So if it were another constant we were comparing with `NSMALLPOSINTS`, let's call it `UCONST` for the sake of the example, would we have to add a cast to `UCONST`?:

```c
do if ((int)UCONST < NSMALLPOSINTS) { \
```

Reply:
> Thanks for the detailed explanations, @mdickinson and @gnprice! I've recently been trying to help out with more PR reviews (since that seems to a bottleneck for CPython), but my experience is primarily with Python and writing documentation, not with C. I have some idea of the fundamentals, but I'm still learning.

Great! I agree that seems to be an important bottleneck.


> So if it were instead an unsigned constant we were comparing with `NSMALLPOSINTS`, let's call it `UCONST` for the sake of the example, would we have to add a cast to `UCONST`?
> 
> ```c
> do if ((int)UCONST < NSMALLPOSINTS) { \
> ```

Pretty much! Just to make the setup explicit -- the case I'm referring to is that there's something like
```
static const unsigned int UCONST = 42;
```
somewhere, perhaps at global scope. Which also means it's not a macro, so we'd typically spell it lowercase:
```
static const unsigned int some_value = 42;
```

If you follow the https://godbolt.org/z/ZYt9pI link above, the example I'm referring to there looks like
```
static const int32_t maxval = 257;

static int
check(uint32_t ival)
{
    return /* ... */ (ival < maxval) /* ... */;
}
```

Then a reasonable response to the warning could be to add that cast, just as you showed, which in this example would make
```
    return /* ... */ (ival < (uint32_t)maxval) /* ... */;
```

Or it might instead be to change the declared type of that const variable:
```
static const uint32_t maxval = 257;
```
Which one to do would depend on which one we felt better matched the way we wanted to think about this constant, and how else it was used.

The third possible response to the warning would be... to notice that there was a bug, and fix the code! That possibility is why the warning exists, after all. :wink: In this example, perhaps that would look like noticing that `ival` really should be signed, and that in fact it comes from someplace where we'd inappropriately cast it from unsigned to signed -- so that if the original unsigned value is large, we'd treat it as negative when we really did want to treat it as large.

Reply:
@gnprice Thank you very much for the detailed explanation! That compiler explorer tool looks incredibly useful for exploring errors (and c in general). The visual highlighting where it shows the c line and it's x86 equivalent is also pretty nice.